csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs

```diff
@@ -47,19 +41,126 @@ public CheckpointState(string checkpointPath)
+        /// Adds the given float property to the checkpoint state.
+        /// <param name="propertyName">Unique name of the property being added.</param>
+        /// <param name="propertyValue">Property value associated with the given name.</param>
+        /// </summary>
+        public void AddProperty(string propertyName, float propertyValue)
         {
-            NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtSaveCheckpoint(handle, NativeOnnxValueHelper.GetPlatformSerializedString(checkpointPath), includeOptimizerState));
+            var propertyNameUtf8 = NativeOnnxValueHelper.StringToZeroTerminatedUtf8(propertyName);
+            float[] value = new float[1];
+            value[0] = propertyValue;
+            Memory<float> memory = value;
+            using (var memHandle = memory.Pin())
+            {
+                IntPtr memPtr;
+                unsafe
+                {
+                    memPtr = (IntPtr)memHandle.Pointer;
+                }
+                NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtAddProperty(handle, propertyNameUtf8, (long)PropertyType.Float, memPtr));
+            }
+        }
+
+        /// <summary>
+        /// Adds the given string property to the checkpoint state.
+        /// <param name="propertyName">Unique name of the property being added.</param>
+        /// <param name="propertyValue">Property value associated with the given name.</param>
+        /// </summary>
+        public void AddProperty(string propertyName, string propertyValue)
+        {
+            var propertyNameUtf8 = NativeOnnxValueHelper.StringToZeroTerminatedUtf8(propertyName);
+            var propertyValueUtf8 = NativeOnnxValueHelper.StringToZeroTerminatedUtf8(propertyValue);
+
+            IntPtr unmanagedPointer = Marshal.AllocHGlobal(propertyValueUtf8.Length);
+            try
+            {
+                Marshal.Copy(propertyValueUtf8, 0, unmanagedPointer, propertyValueUtf8.Length);
+                NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtAddProperty(handle, propertyNameUtf8, (long)PropertyType.String, unmanagedPointer));
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(unmanagedPointer);
+            }
+        }
+
+        /// <summary>
+        /// Gets the property value associated with the given name from the checkpoint state.
+        /// <param name="propertyName">Unique name of the property being retrieved.</param>
+        /// </summary>
+        public object GetProperty(string propertyName)
+        {
+            var propertyNameUtf8 = NativeOnnxValueHelper.StringToZeroTerminatedUtf8(propertyName);
+            var allocator = OrtAllocator.DefaultInstance;
+            IntPtr propertyValue = IntPtr.Zero;
+            NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtGetProperty(handle, propertyNameUtf8, allocator.Pointer, out long propertyType, out propertyValue));
+
+            if (propertyType == (long)PropertyType.Int)
```

Comment:
nit: EnumTryParse propertyType first and then do comparisons with PropertyType without casts.

Reply:
Can get away with needing to do this since I can now get the `propertyType` as `PropertyType`.